### DISCUSSION TITLE
Plex support ?

### DISCUSSION THREAD

**Discussion by WoisWoi** [qdm12/gluetun#525] [Q&A]
Hi :)
I tried to use it for my plex media server (ubuntu server) with PIA and port forwarding, but remote access to the server is not working.
Any idea how to make this work ?

**Comment by qdm12** [comment]
PIA's port forwarding doesn't really work somehow for most applications, although it works for torrenting over udp I think.

You can test it out with https://github.com/qdm12/gluetun/wiki/Port-forwarding#test-it (spin up an http server on gluetun).

The same works for both Mullvad and VyprVPN AFAIK, and I have some users who reported the PIA bug also occurs on native openvpn without docker/gluetun.

I would suggest you Reach out to PIA's support about it, I'd be curious why it is like so. Also https://github.com/qdm12/gluetun/issues/464 might interest you, I will write a wiki page about it soon.

**Reply by WoisWoi** [reply]
> 
> 
> PIA's port forwarding doesn't really work somehow for most applications, although it works for torrenting over udp I think.
> 
> You can test it out with https://github.com/qdm12/gluetun/wiki/Port-forwarding#test-it (spin up an http server on gluetun).
> 
> The same works for both Mullvad and VyprVPN AFAIK, and I have some users who reported the PIA bug also occurs on native openvpn without docker/gluetun.
> 
> I would suggest you Reach out to PIA's support about it, I'd be curious why it is like so. Also #464 might interest you, I will write a wiki page about it soon.

OK, I'll check it out tomorrow. Thanks you.
Then is there a better VPN provider for a working port forwarding ?
Also, do you know is there is any VPN provider really capable to follow up a server with a 10 Gigabit bandwidth ?

**Reply by qdm12** [reply]
> Then is there a better VPN provider for a working port forwarding

As far as I know, Mullvad works well today and VyprVPN worked well a year ago for me. PIA never worked for me for years and I dropped it out a year ago. Not sure about others though.

> Also, do you know is there is any VPN provider really capable to follow up a server with a 10 Gigabit bandwidth

That depends on your client location and vpn server location. The furthest it is the smaller the bandwidth will be usually because there are bottlenecks on the route from client to server.
Then it depends on the load of the VPN server which varies with time.
For over 1gbps, it also depends on the client cpu, especially encryption hardware acceleration, so make sure you're good on that.
So a lot of factors to take in, I can't give you a precise answer for that and never really benched them since I don't need that much bandwidth anyway 😉

**Reply by WoisWoi** [reply]
![image](https://user-images.githubusercontent.com/75491231/126303855-ccd910bc-fbc5-444c-bfc7-efaa4a04aedd.png)
I'm trying VyprVPN right now, shall I setup something else ? It's still not working for me :(

Edit : 
![image](https://user-images.githubusercontent.com/75491231/126304540-512ee62c-9847-41ba-96b3-00eb5a12344f.png)

**Reply by qdm12** [reply]
See this https://github.com/qdm12/gluetun/wiki/Port-forwarding 

How are you testing it? Make sure you make plex listen on the forwarded port and connect to it using vpn-ip:forwarded-port

EDIT: While it is not possible to open specific inbound ports while connected to VyprVPN, it is possible to open all inbound ports, which can be done by disabling NAT Firewall in the Manage Services section of your Account. (source: https://support.vyprvpn.com/hc/en-us/articles/360039668472-Does-VyprVPN-support-Port-Forwarding-)

**Reply by WoisWoi** [reply]
Je viens de voir que tu parlais français :)
J'ai bien sélectionné manuellement le port dans plex, dans la section "accès à distance".

Donc il faut que je désactives le firewall, ce n'est pas possible de "port forwardé" un port directement, même désigné aléatoirement ?

Edit : Mullvad le permet ?
![image](https://user-images.githubusercontent.com/75491231/126339628-fd766b81-7e0e-4da4-ab7a-c56723fc4e51.png)
![image](https://user-images.githubusercontent.com/75491231/126339809-bbd17194-5520-42d0-b572-aef6f9b2aa7b.png)
Je viens de désactiver le NAT Firewall comme indiqué, j'ai les résultats visibles en screenshot. Peut-être qu'il y a un délai pour la prise d'effet ?

**Reply by qdm12** [reply]
Oui je parle francais (sans accents 😉)

Est-ce que tu confondrais peut-etre server side vpn port forwarding avec port mapping.

Qu'essaies-tu de faire?

- Si tu veux acceder a tonServeurIP:32400, vas voir [cette page Wiki](https://github.com/qdm12/gluetun/wiki/Connect-to-gluetun#access-ports-of-containers-connected-to-gluetun)
- Si tu veux acceder a vpnIP:32400, pas besoin de port map avec Docker. Il faut activer ca sur ton compte vyprvpn ([voir ici](https://support.vyprvpn.com/hc/en-us/articles/360039668472-Does-VyprVPN-support-Port-Forwarding-)) et mettre dans gluetun la variable `FIREWALL_VPN_INPUT_PORTS=32400`. A ce moment tu devrais pouvoir te connecter a `vpn-server-ip:32400`. Tu peux trouver ton adresse ip vpn dans les logs de gluetun par exemple.

**Reply by WoisWoi** [reply]
Je veux cacher l'adresse de ce serveur Plex pour les utilisateurs derrière un VPN. 
Donc  vpnIP:32400 ou un autre port (parce que j'ai déjà le port 32400 utilisé par un serveur serveur plex qui n'est pas dans un container), idem via l'app plex :)

Et du coup, ça m'a l'air bien configuré : 

port remappé dans portainer sur gluetun -> 
![image](https://user-images.githubusercontent.com/75491231/126379841-a8bed73e-b388-4700-b6a2-e1b69db53dad.png)

la variable d'environnement FIREWALL_VPN_INPUT_PORTS -> 
![image](https://user-images.githubusercontent.com/75491231/126379906-1e7d6095-c846-44ef-81c4-1c5a0b359bed.png)

et j'ai désactivé le NAT firewall sur mon compte VyprVPN
Et malgré tout, Plex n'arrive pas à se connecter à internet via le port 32450 ou 32400

**Reply by IMgoRt** [reply]
> Je viens de voir que tu parlais français :) J'ai bien sélectionné manuellement le port dans plex, dans la section "accès à distance".
> 
> Donc il faut que je désactives le firewall, ce n'est pas possible de "port forwardé" un port directement, même désigné aléatoirement ?
> 
> Edit : Mullvad le permet ? ![image](https://user-images.githubusercontent.com/75491231/126339628-fd766b81-7e0e-4da4-ab7a-c56723fc4e51.png) ![image](https://user-images.githubusercontent.com/75491231/126339809-bbd17194-5520-42d0-b572-aef6f9b2aa7b.png) Je viens de désactiver le NAT Firewall comme indiqué, j'ai les résultats visibles en screenshot. Peut-être qu'il y a un délai pour la prise d'effet ?

For me I added both the standard plex port and the VPN port to the environment variable (in your example):

`FIREWALL_VPN_INPUT_PORTS: 32400,32450 `

**Comment by qdm12** [comment]
Essaies d'abord en suivant les instructions [ici](https://github.com/qdm12/gluetun/wiki/Port-forwarding#test-it) et en testant le port 32400, pour etre sur que le port est accessible. Petite note, pour moi le port 32400 suffit pour tout.

Tu peux aussi essayer de desactiver le firewall de gluetun `FIREWALL=off` pour tester (mais pas le garder comme ca car sinon ca leak de partout).

**Reply by WoisWoi** [reply]
![image](https://user-images.githubusercontent.com/75491231/126385757-50aa8cd7-0572-4e08-bd5c-b9a8d1a47769.png)
![image](https://user-images.githubusercontent.com/75491231/126385985-14d63dce-4e2e-4a8d-a675-5e94a38da88c.png)
(gluetun FIREWALL was on)

-> now with the gluetun FIREWALL off 
![image](https://user-images.githubusercontent.com/75491231/126386411-da1a0113-c890-4f09-852d-b09142e2e0d0.png)
![image](https://user-images.githubusercontent.com/75491231/126386624-c019c2bf-e115-4774-a871-a809de84d5d9.png)

_Edit : j'ai peur de créer un conflit si j'utilise le port 32400 comme j'ai déjà un serveur Plex en fonctionnement sur ce port et sur le même serveur, mais qui n'est pas dans un container._

**Reply by qdm12** [reply]
Ok Cool donc le port forward marche. Ne te fie pas trop a la croix rouge, essaie directement d'y acceder avec ton browser a `<vpn-ip>:32450` pour voir?

Moi aussi j'ai une croix rouge (bien que je ne passe pas par gluetun):

![image](https://user-images.githubusercontent.com/15654980/126391780-8aeb2578-9058-47b8-bd0c-a568b3a327e7.png)

Mais j'y ai quand meme access sans probleme (via un reverse proxy sur mon serveur).

> j'ai peur de créer un conflit si j'utilise le port 32400 comme j'ai déjà un serveur Plex en fonctionnement sur ce port et sur le même serveur, mais qui n'est pas dans un container.

Tu ne peux pas faire de conflit vu que ce plex est dans le vpn, a moins que tu port map 32400:32400 dans Portainer pour gluetun.
Et meme si tu voulais lancer un autre plex sur port 32400, vu que tu as deja une une application plex sur 32400, c'est juste la nouvelle qui ne lancera pas et te donnera une erreur, donc il y a aucun moyen d'avoir un conflit 😉 Tu ne peux pas avoir 2 processes sur le meme port.

**Reply by qdm12** [reply]
Aussi si port checker marche avec `FIREWALL=on`, c'est que gluetun marche comme il faut, donc tu peux laisser le firewall on.

Et une autre note, n'oublie pas que vu que tu n'utilise pas https pour plex, tu es dans la situation

```
plex + gluetun --encrypte--> VPN server (pas encrypte) --pas encrypte--> clients
```

Donc le serveur VPN peut voir l'information et sniffers entre le serveur VPN et les clients peuvent aussi voir.

Si tu utilisais https avec un reverse proxy comme caddy par exemple tu pourrais avoir:

```
plex + gluetun + caddy --encrypte x2--> VPN server (encrypte) --encrypte --> clients
```

Voir [ca](https://github.com/qdm12/gluetun/issues/464#issuecomment-848063080) pour comment faire. C'est encore approximatif, mais je compte ecrire une page wiki plus claire bientot.

**Reply by WoisWoi** [reply]
Je fais déjà du reverse proxy avec nginx proxy manager (container), mais je n'ai rien configuré du coup pour le vpn

edit : 
![image](https://user-images.githubusercontent.com/75491231/126393342-338234bb-57a4-40e2-9359-c309e4265399.png)
![image](https://user-images.githubusercontent.com/75491231/126393380-8520e9e3-5058-4494-8798-441720bf248d.png)

J'ai retesté, le port forwading fonctionne bien, mais c'est comme si Plex n'arrivait pas à l'utiliser, je comprends pas trop ce qui cloche

**Reply by qdm12** [reply]
Ah je sais, c'est car plex ecoute sur 32400, il faut que changes ca a 32450. Pas sur comment faire, et meme pas sur que ce soit possible de changer le port interne.

Sinon mets un reverse proxy (nginx ou autre) sur le port 32450 qui redirige `/` vers `http://localhost:32400` et ca devrait marcher.
Comme ca tu peux meme setup tls plus tard si tu veux.

EDIT: ou tu peux le faire avec `iptables` dans gluetun, voir https://github.com/qdm12/gluetun/wiki/Advanced-setup#custom-iptables-rules et rediriger tout ce qui vient sur 32450 vers 32400.

EDIT 2: Je te conseille de just mettre 32400 dans le port externe dans plex et y acceder via vpnIP:32400 ca ne conflictera pas 😉 Dans un premier temps

**Reply by WoisWoi** [reply]
![image](https://user-images.githubusercontent.com/75491231/126407661-dcebba4a-cdb9-47a8-a0bb-4ce6bc830d5b.png)
![image](https://user-images.githubusercontent.com/75491231/126407683-13cf063e-b9b7-4e8d-9d93-03b5ea1267e8.png)

Même si c'est gluetun, ça passe visiblement pas 
En tout cas merci pour ton temps :)

**Reply by qdm12** [reply]
Si juste change le dans le port externe dans l'ui plex. Tu peux laisser le port mapping de Portainer à 32450:32000.

**Reply by WoisWoi** [reply]
ça a l'air de fonctionner, merci 👍🏼 Je creuserais ça demain :)

**Reply by WoisWoi** [reply]
> 
> 
> Si juste change le dans le port externe dans l'ui plex. Tu peux laisser le port mapping de Portainer à 32450:32000.

Je viens de faire des tests et je constate des problèmes de débit curieux :

Serveur Plex sous VPN sur le serveur NL = bottleneck /oscille entre 15 et 50 mbps, environ ~ 20 mbps en lecture (débit que je reçois sur mon ordi en lecture*)

Si je test ce même serveur VPN depuis mon ordi avec l'appli windows j'ai 450 mbps symétrique au speedtest et si je test sur mon autre serveur plex qui n'est pas sous VPN en étant connecté avec l'appli sur mon ordi du coup sur le serveur NL, je ne constate pas de problème de débit, ça m'envoie ~330 Mbps comme d'habitude. Qu'est-ce qui peut provoquer ça ?

**Reply by qdm12** [reply]
Ca peut etre

1. Docker qui bottleneck si ton host est une VM/Windows/OSX
2. Vyprvpn qui limitent le debit sortant du serveur VPN pour X raison (faut leur demander)
3. Voir [cette page](https://github.com/qdm12/gluetun/wiki/Bandwidth-speeds)

Est-ce que tu utilise le meme CPU pour ton plex sans vpn et tes containers?

**Reply by WoisWoi** [reply]
> 
> 
> Ca peut etre
> 
>     1. Docker qui bottleneck si ton host est une VM/Windows/OSX
> 
>     2. Vyprvpn qui limitent le debit sortant du serveur VPN pour X raison (faut leur demander)
> 
>     3. Voir [cette page](https://github.com/qdm12/gluetun/wiki/Bandwidth-speeds)
> 
> 
> Est-ce que tu utilise le meme CPU pour ton plex sans vpn et tes containers?

Oui, même CPU, mais mon serveur est une machine de guerre (du coup pour te répondre c'est pas une VM, c'est un dédié sous linux).
AMD EPYC 7502P, 128 Go RAM DDR4 ECC, 4 to SSD NVMe, 10 Gbit symétrique

P.S : the server is in Helsinki and I'm in France

**Reply by qdm12** [reply]
As-tu essaye sur d'autres serveurs? C'est vrai que c'est etrange, essaies de faire un speedtest en tant que client dans gluetun.

Par exemple:

```sh
docker exec gluetun apk add speedtest-cli && speedtest-cli
```

**Reply by WoisWoi** [reply]
> 
> 
> As-tu essaye sur d'autres serveurs? C'est vrai que c'est etrange, essaies de faire un speedtest en tant que client dans gluetun.
> 
> Par exemple:
> 
> ```shell
> docker exec gluetun apk add speedtest-cli && speedtest-cli
> ```

![image](https://user-images.githubusercontent.com/75491231/126560770-d35b1431-797f-408b-a0e9-a5d5ce24f246.png)
je vais essayer un autre serveur 

Edit : le problème pour tester d'autres serveurs, c'est que la variable "hostname" n'a pas l'air de bien fonctionner. Elle a fonctionné directement pour le serveur au Pays-bas, mais pour tous les autres, j'ai l'impression que le paramètre n'est pas pris en compte. Je continue de tester d'autres serveurs...

Autre speedtest depuis un serveur VPN à Londres, je n'arrive pas à lui imposer un serveur...
![image](https://user-images.githubusercontent.com/75491231/126567175-05d6ba91-e0bb-4e5c-a19b-ca6f770ed6e7.png)

**Reply by WoisWoi** [reply]
> 
> 
> As-tu essaye sur d'autres serveurs? C'est vrai que c'est etrange, essaies de faire un speedtest en tant que client dans gluetun.
> 
> Par exemple:
> 
> ```shell
> docker exec gluetun apk add speedtest-cli && speedtest-cli
> ```

Pas d'idée de ce que ça pourrait être ?

**Reply by qdm12** [reply]
J'ai retravaillé ça récemment si ça peut t'intéresser:

https://github.com/qdm12/gluetun/wiki/Bandwidth-speeds

TLDR Docker rajoute de l'overhead, machine de guerre ou pas.
